File: data_structures/union_find/solutions.py
"""Union-Find — disjoint set with path compression and union by rank."""

from typing import List


class UnionFind:
    """Disjoint set. Find/union are effectively O(1) amortized."""

    def __init__(self, n: int) -> None:
        self.parent = list(range(n))
        self.rank = [0] * n
        self.components = n

    def find(self, x: int) -> int:
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, a: int, b: int) -> bool:
        """Merge groups. Returns False if a and b were already connected."""
        ra, rb = self.find(a), self.find(b)
        if ra == rb:
            return False
        if self.rank[ra] < self.rank[rb]:
            ra, rb = rb, ra
        self.parent[rb] = ra
        if self.rank[ra] == self.rank[rb]:
            self.rank[ra] += 1
        self.components -= 1
        return True

    def connected(self, a: int, b: int) -> bool:
        return self.find(a) == self.find(b)


def find_circle_num(is_connected: List[List[int]]) -> int:
    """Number of Provinces — friend circles in an n×n matrix. O(n²)."""
    n = len(is_connected)
    uf = UnionFind(n)
    for i in range(n):
        for j in range(i + 1, n):
            if is_connected[i][j] == 1:
                uf.union(i, j)
    return uf.components


def valid_tree(n: int, edges: List[List[int]]) -> bool:
    """Graph Valid Tree — n-1 edges, no cycle, one component. O(n)."""
    if len(edges) != n - 1:
        return False
    uf = UnionFind(n)
    for a, b in edges:
        if not uf.union(a, b):
            return False
    return uf.components == 1


def equations_possible(equations: List[str]) -> bool:
    """Satisfiability of Equality Equations over letters a–z. O(n)."""
    uf = UnionFind(26)
    for eq in equations:
        if eq[1] == "=":
            uf.union(ord(eq[0]) - ord("a"), ord(eq[3]) - ord("a"))
    for eq in equations:
        if eq[1] == "!":
            a, b = ord(eq[0]) - ord("a"), ord(eq[3]) - ord("a")
            if uf.connected(a, b):
                return False
    return True

File: data_structures/union_find/Solutions.java
import java.util.*;

/** Union-Find — disjoint set with path compression and union by rank. */
public class Solutions {

    /** Disjoint set. Find/union are effectively O(1) amortized. */
    public static class UnionFind {
        private final int[] parent;
        private final int[] rank;
        private int components;

        public UnionFind(int n) {
            parent = new int[n];
            rank = new int[n];
            components = n;
            for (int i = 0; i < n; i++) parent[i] = i;
        }

        public int find(int x) {
            if (parent[x] != x) parent[x] = find(parent[x]);
            return parent[x];
        }

        /** Merge groups. Returns false if a and b were already connected. */
        public boolean union(int a, int b) {
            int ra = find(a), rb = find(b);
            if (ra == rb) return false;
            if (rank[ra] < rank[rb]) {
                int tmp = ra;
                ra = rb;
                rb = tmp;
            }
            parent[rb] = ra;
            if (rank[ra] == rank[rb]) rank[ra]++;
            components--;
            return true;
        }

        public boolean connected(int a, int b) {
            return find(a) == find(b);
        }

        public int getComponents() {
            return components;
        }
    }

    /** Number of Provinces — friend circles in an n×n matrix. O(n²). */
    public static int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;
        UnionFind uf = new UnionFind(n);
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                if (isConnected[i][j] == 1) uf.union(i, j);
            }
        }
        return uf.getComponents();
    }

    /** Graph Valid Tree — n-1 edges, no cycle, one component. O(n). */
    public static boolean validTree(int n, int[][] edges) {
        if (edges.length != n - 1) return false;
        UnionFind uf = new UnionFind(n);
        for (int[] edge : edges) {
            if (!uf.union(edge[0], edge[1])) return false;
        }
        return uf.getComponents() == 1;
    }

    /** Satisfiability of Equality Equations over letters a–z. O(n). */
    public static boolean equationsPossible(String[] equations) {
        UnionFind uf = new UnionFind(26);
        for (String eq : equations) {
            if (eq.charAt(1) == '=') {
                uf.union(eq.charAt(0) - 'a', eq.charAt(3) - 'a');
            }
        }
        for (String eq : equations) {
            if (eq.charAt(1) == '!') {
                if (uf.connected(eq.charAt(0) - 'a', eq.charAt(3) - 'a')) return false;
            }
        }
        return true;
    }
}
